File: src/236. Lowest Common Ancestor of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    public:
        TreeNode * lowestCommonAncestor(TreeNode * root, TreeNode * p, TreeNode * q) {
            if (root == NULL || root == p || root == q) {
                return root;
            }
            TreeNode * lcaL = lowestCommonAncestor(root -> left, p, q);
            TreeNode * lcaR = lowestCommonAncestor(root -> right, p, q);
            if (lcaL && lcaR) {
                return root;
            }
            if (lcaL) {
                return lcaL;
            }
            return lcaR;
        }
};

File: src/236. Lowest Common Ancestor of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not root or root is p or root is q:
            return root
        lcaR = self.lowestCommonAncestor(root.left, p, q)
        lcaL = self.lowestCommonAncestor(root.right, p, q)
        return root if lcaR and lcaL else lcaR or lcaL
        